src/cli/cli_coap.cpp

```diff
@@ -0,0 +1,356 @@
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a simple CLI coap server and client.
+ */
+
+#include <cli/cli.hpp>
+#include <cli/cli_coap.hpp>
+
+namespace Thread {
+namespace Cli {
+
+const CoapCommand Coap::sCommands[] =
+{
+    { "client", &ProcessClient },
+    { "server", &ProcessServer }
+};
+
+Server *Coap::sServer;
+otCoapResource *Coap::sResource;
+otInstance *Coap::sInstance;
+
+void Coap::PrintPayload(otMessage *aMessage)
+{
+    uint8_t payloadLength = otMessageGetLength(aMessage) - otMessageGetOffset(aMessage);
+    uint8_t outputBufferLength = 16;
+    uint8_t bytesPrinted = 0;
+
+    if (payloadLength > 0)
+    {
+        char payload[outputBufferLength] = {};
+        while (bytesPrinted < payloadLength)
+        {
+            otMessageRead(aMessage, otMessageGetOffset(aMessage) + bytesPrinted, &payload, outputBufferLength);
+            bytesPrinted = bytesPrinted + outputBufferLength;
+            sServer->OutputFormat(" payload: %s", payload);
+        }
+    }
+    sServer->OutputFormat("\r\n");
+}
+
+ThreadError Coap::Process(otInstance *aInstance, int argc, char *argv[], Server &aServer, otCoapResource &aResource)
+{
+    ThreadError error = kThreadError_None;
+
+    sInstance = aInstance;
+    sServer = &aServer;
+    sResource = &aResource;
+    sResource->mHandler = (otCoapRequestHandler) &Coap::s_HandleServerResponse;
+
+    if (argc == 0)
+    {
+        ExitNow(error = kThreadError_InvalidArgs);
+    }
+
+    for (unsigned int i = 0; i < sizeof(sCommands) / sizeof(sCommands[0]); i++)
+    {
+        if (strcmp(argv[0], sCommands[i].mName) == 0)
+        {
+            error = sCommands[i].mCommand(argc - 1, argv + 1);
+            break;
+        }
+    }
+
+exit:
+    return error;
+}
+
+ThreadError Coap::ProcessServer(int argc, char *argv[])
+{
+    ThreadError error = kThreadError_None;
+
+    VerifyOrExit(argc > 0, error = kThreadError_InvalidArgs);
+
+    if (strcmp(argv[0], "start") == 0)
+    {
+        SuccessOrExit(error = otCoapServerStart(sInstance));
+        SuccessOrExit(error = otCoapServerAddResource(sInstance, sResource));
+        sServer->OutputFormat("Server started with resource '%s': ", sResource->mUriPath);
+    }
+    else if (strcmp(argv[0], "stop") == 0)
+    {
+        otCoapServerRemoveResource(sInstance, sResource);
+        SuccessOrExit(error = otCoapServerStop(sInstance));
+        sServer->OutputFormat("Server stopped: ");
+    }
+    else if (strcmp(argv[0], "name") == 0)
+    {
+        if (argc > 1)
+        {
+            sResource->mUriPath = argv[1];
```

Comment:
This still will not work. You need to declare some buffer for Uri Path and assign it to mUriPath. Then if you want to change name - change content of the buffer.